bundle/src/main/java/com/adobe/acs/commons/fetchfile/impl/FileFetchImpl.java

```diff
@@ -0,0 +1,229 @@
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.sling.api.resource.LoginException;
+import org.apache.sling.api.resource.ModifiableValueMap;
+import org.apache.sling.api.resource.PersistenceException;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.xmlbeans.impl.util.Base64;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.metatype.annotations.Designate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.adobe.acs.commons.fetchfile.FileFetch;
+import com.adobe.acs.commons.fetchfile.FileFetchConfiguration;
+import com.day.cq.dam.api.Asset;
+import com.day.cq.dam.api.AssetManager;
+import com.day.cq.replication.ReplicationActionType;
+import com.day.cq.replication.ReplicationException;
+import com.day.cq.replication.Replicator;
+
+/**
+ * Implementation of the FileFetch service
+ */
+@Component(service = { Runnable.class, FileFetch.class }, configurationPolicy = ConfigurationPolicy.REQUIRE)
+@Designate(ocd = FileFetchConfiguration.class, factory = true)
+public class FileFetchImpl implements FileFetch, Runnable {
+
+  private static final Logger log = LoggerFactory.getLogger(FileFetchImpl.class);
+
+  private static final String SERVICE_USER_NAME = "file-fetch";
+
+  protected FileFetchConfiguration config;
+
+  @Reference
+  private ResourceResolverFactory factory;
+
+  private Exception lastException = null;
+
+  private boolean lastJobSucceeded = true;
+
+  @Reference
+  private Replicator replicator;
+
+  @Activate
+  public void activate(FileFetchConfiguration config)  {
+    log.info("Activing with configuration: {}", config);
+    this.config = config;
+    run();
+  }
+
+  private byte[] fetchFile() throws IOException {
+    log.trace("fetchFile");
+
+    HttpURLConnection con = null;
+    try {
+      log.debug("Opening connection to {}", config.remoteUrl());
+      con = openConnection();
+
+      List<Pair<String, String>> header = Arrays.stream(config.headers()).map(h -> h.split("\\="))
+          .filter(h -> (h.length >= 2)).map(h -> {
+            String value = Arrays.stream(ArrayUtils.remove(h, 0)).collect(Collectors.joining("="));
+            return new ImmutablePair<String, String>(h[0], value);
+          }).collect(Collectors.toList());
+      for (Pair<String, String> p : header) {
+        log.trace("Adding request property {}={}", p.getKey(), p.getValue());
+        con.addRequestProperty(p.getKey(), p.getValue());
+      }
+      con.setUseCaches(false);
+      int rc = con.getResponseCode();
+      log.debug("Retrieved response code: {}", rc);
+      if (ArrayUtils.contains(config.validResponseCodes(), rc)) {
```

Comment:
Wouldn't it make sense to check the statuscode or even send an "if-modified-since" header to avoid downloading the file alltogether if it hasn't changed?

Reply:
Ah yes, this is a much better idea and allows for avoiding the large binary issue above since I no longer have to compute the hash.